In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/IDRA/Day16_Student_Wellbeing_Survey.csv")
df.head()

,Student_ID,Age,Faculty,Year_of_Study,City,Accommodation,Scholarship,Part_Time_Job,Internet_Quality,Preferred_Study_Space,Weekly_Study_Hours,Average_Sleep_Hours,Daily_Screen_Time_Hours,Exercise_Days_Per_Week,Commute_Time_Minutes,Stress_Score,Academic_Readiness_Score,Overall_Satisfaction,Monthly_Discretionary_Spending,Social_Activity_Hours_Per_Week
0,STU0001,21,Data Science,3,Bengaluru,Hostel,No,Yes,Good,Café,15.2,5.7,3.2,3,11.8,5.4,63.1,3.2,5131.0,7.8
1,STU0002,21,Data Science,4,Pune,Hostel,Yes,No,Poor,Library,17.6,7.7,3.5,4,20.3,2.1,74.2,3.8,3508.0,9.2
2,STU0003,22,Life Sciences,2,Chandigarh,Home,Yes,No,Poor,Home,18.9,6.4,2.8,1,15.0,5.2,77.3,4.0,4584.0,7.7
3,STU0004,18,Business,4,Bengaluru,Hostel,Yes,Yes,Average,Café,19.5,6.7,3.0,5,12.1,4.7,67.0,4.0,5456.0,5.3
4,STU0005,20,Business,3,Jammu,Shared Apartment,Yes,No,Good,Home,18.4,7.3,3.0,1,30.2,5.4,65.6,3.2,11732.0,15.9


In [3]:
df.shape

(600, 20)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 600 entries, 0 to 599
Data columns (total 20 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Student_ID                      600 non-null    object 
 1   Age                             600 non-null    int64  
 2   Faculty                         600 non-null    object 
 3   Year_of_Study                   600 non-null    int64  
 4   City                            600 non-null    object 
 5   Accommodation                   600 non-null    object 
 6   Scholarship                     600 non-null    object 
 7   Part_Time_Job                   600 non-null    object 
 8   Internet_Quality                600 non-null    object 
 9   Preferred_Study_Space           600 non-null    object 
 10  Weekly_Study_Hours              600 non-null    float64
 11  Average_Sleep_Hours             600 non-null    float64
 12  Daily_Screen_Time_Hours         600 

In [5]:
df.isnull().sum().sum()

np.int64(0)

In [6]:
df.duplicated().sum()

np.int64(0)

#Mean, Median, Mode of Columns

In [7]:
m_cols = ['Weekly_Study_Hours', 'Average_Sleep_Hours', 'Daily_Screen_Time_Hours', 'Stress_Score', 'Academic_Readiness_Score']

stats_data = []
for col in m_cols:
  mean_val = df[col].mean()
  median_val = df[col].median()
  mode_val = df[col].mode()[0]
  stats_data.append({'Column': col, 'Mean': mean_val, 'Median': median_val, 'Mode': mode_val})

stats_df = pd.DataFrame(stats_data)
display(stats_df)

,Column,Mean,Median,Mode
0,Weekly_Study_Hours,15.691000,15.40,13.1
1,Average_Sleep_Hours,6.997500,7.00,7.0
2,Daily_Screen_Time_Hours,4.503000,4.20,3.5
3,Stress_Score,4.464500,4.50,4.7
4,Academic_Readiness_Score,71.773167,71.65,71.1


##Formulae used
Mean = sum of observations / no. of observations

Median = middle value after sorting the observations

Mode = most frequently occuring observation value

#Range, Variance, Standard deviation, Q1, Q3, IQR

In [8]:
from numpy._core.fromnumeric import var
measures_data = []
for col in m_cols:
  range_val = df[col].max() - df[col].min()
  var_val = df[col].var()
  std_val = df[col].std()
  q1_val = df[col].quantile(0.25)
  q3_val = df[col].quantile(0.75)
  iqr_val = q3_val - q1_val
  measures_data.append({'Column': col, 'Range': range_val, 'Variance': var_val, 'Standard Deviation': std_val, 'Q1': q1_val, 'Q3': q3_val, 'IQR': iqr_val})

measures_df = pd.DataFrame(measures_data)
display(measures_df)

,Column,Range,Variance,Standard Deviation,Q1,Q3,IQR
0,Weekly_Study_Hours,30.0,19.217047,4.383725,13.000,18.425,5.425
1,Average_Sleep_Hours,5.0,0.703383,0.838679,6.475,7.600,1.125
2,Daily_Screen_Time_Hours,11.2,3.469006,1.862527,3.200,5.400,2.200
3,Stress_Score,8.4,2.782494,1.668081,3.300,5.700,2.400
4,Academic_Readiness_Score,56.1,93.981967,9.694430,65.200,78.025,12.825


In [9]:
print("\nVariable with greatest variability:", df[m_cols].std().idxmax())


Variable with greatest variability: Academic_Readiness_Score


The variable having the greatest variability can be found by comparing the variance of different columns or finding the largest standard deviation.

In this case, by observation, it could be understood that the variable with the greatest variability is **Academic_Readiness_Score**

#Outlier Detection using IQR

In [10]:
outlier_col = ['Weekly_Study_Hours', 'Daily_Screen_Time_Hours', 'Commute_Time_Minutes', 'Monthly_Discretionary_Spending']

In [11]:
outlier_info = []
for col in outlier_col:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outliers = df[(df[col] < lower_bound) | (df[col] > upper_bound)]

    outlier_info.append({
        'Column': col,
        'Q1': Q1,
        'Q3': Q3,
        'IQR': IQR,
        'Lower Bound': lower_bound,
        'Upper Bound': upper_bound,
        'Number of Outliers': len(outliers)
    })

outlier_df = pd.DataFrame(outlier_info)
print(outlier_df)

                           Column        Q1        Q3       IQR  Lower Bound  \
0              Weekly_Study_Hours    13.000    18.425     5.425       4.8625   
1         Daily_Screen_Time_Hours     3.200     5.400     2.200      -0.1000   
2            Commute_Time_Minutes    13.375    31.525    18.150     -13.8500   
3  Monthly_Discretionary_Spending  4113.000  7808.500  3695.500   -1430.2500   

   Upper Bound  Number of Outliers  
0      26.5625                   8  
1       8.7000                  18  
2      58.7500                   4  
3   13351.7500                  20  


In [12]:
outlier_stats = []

for col in outlier_col:
  mean_before = df[col].mean()
  median_before = df[col].median()

  q1 = df[col].quantile(0.25)
  q3 = df[col].quantile(0.75)
  iqr = q3 - q1
  low_bound = q1 - 1.5 * iqr
  high_bound = q3 + 1.5 * iqr

  df_filtered_outliers = df[(df[col] >= low_bound) & (df[col] <= high_bound)]

  mean_after = df_filtered_outliers[col].mean()
  median_after = df_filtered_outliers[col].median()

  outlier_stats.append({
      'Column': col,
      'Lower Bound': low_bound,
      'Upper Bound': high_bound,
      'Mean_Before_Outliers': mean_before,
      'Median_Before_Outliers': median_before,
      'Mean_After_Outliers': mean_after,
      'Median_After_Outliers': median_after
  })

outlier_comparison_df = pd.DataFrame(outlier_stats)
display(outlier_comparison_df)

,Column,Lower Bound,Upper Bound,Mean_Before_Outliers,Median_Before_Outliers,Mean_After_Outliers,Median_After_Outliers
0,Weekly_Study_Hours,4.8625,26.5625,15.691000,15.40,15.602872,15.35
1,Daily_Screen_Time_Hours,-0.1000,8.7000,4.503000,4.20,4.313402,4.10
2,Commute_Time_Minutes,-13.8500,58.7500,22.893833,20.45,22.513926,20.35
3,Monthly_Discretionary_Spending,-1430.2500,13351.7500,6343.778333,5685.50,5973.718966,5590.50


#Probability

In [13]:
A = df['Part_Time_Job'] == 'Yes'
B = df['Stress_Score'] >= 7
C = df['Scholarship'] == 'Yes'
D = df['Exercise_Days_Per_Week'] >= 3

In [14]:
P_A = A.mean()
P_B = B.mean()
P_C = C.mean()
P_D = D.mean()

P_A_and_B = (A & B).mean()
P_A_or_B = P_A + P_B - P_A_and_B

P_A_given_B = P_A_and_B / P_B
P_B_given_A = P_A_and_B / P_A


print("P(A) =", P_A)
print("P(B) =", P_B)
print("P(C) =", P_C)
print("P(D) =", P_D)
print("\n")

print("P(A or B) =", P_A_or_B)
print("P(A and B) =", P_A_and_B)
print("\n")

print("P(A|B) =", P_A_given_B)
print("P(B|A) =", P_B_given_A)

P(A) = 0.25333333333333335
P(B) = 0.075
P(C) = 0.30666666666666664
P(D) = 0.5933333333333334


P(A or B) = 0.2766666666666667
P(A and B) = 0.051666666666666666


P(A|B) = 0.6888888888888889
P(B|A) = 0.2039473684210526


##Formula
P(A) = no. of students satisfying condition A / total no. of students

P(A∪B) = P(A) + P(B) − P(A∩B)

P(A∩B) = no. of students satisfying both A and B / total students

P(A|B) = P(A∩B) / P(B)

#Mutually Exclusive Events

In [15]:
yr_1 = df['Year_of_Study']==1
yr_4 = df['Year_of_Study']==4

P_yr1_and_yr4 = (yr_1 & yr_4).mean()
print(P_yr1_and_yr4)

0.0


The result of their intersection is 0, indicating that they are mutually exclusive.

A student can't be in Year 1 and Year 4 simultaneously, so they are mutually exclusive events

#Independent Events

Formula for independent events: P(A∩B)=P(A)×P(B)

In [16]:
P_A_times_P_B = P_A * P_B

print("P(A and B) =", P_A_and_B)
print("P(A) × P(B) =", P_A_times_P_B)
print("Difference =", abs(P_A_and_B - P_A_times_P_B))

P(A and B) = 0.051666666666666666
P(A) × P(B) = 0.019
Difference = 0.03266666666666666


Since P(A and B) is approximately equal to P(A) × P(B), events A and B appear to be independent.

#Bayes' Theorem

P(A|B) = [P(B|A) * P(A)] / [(P(B|A) * P(A)) + (P(B|**~**A) * P(**~**A))]

In [17]:
P_not_A = 1 - P_A

P_B_given_not_A = (B & ~A).mean() / (~A).mean()
print("P(B|not A) =", P_B_given_not_A)

P_A_given_B_bayes = (
    P_B_given_A * P_A
) / (
    (P_B_given_A * P_A) +
    (P_B_given_not_A * P_not_A)
)
print("P(A|B) using Bayes' theorem =", P_A_given_B_bayes)

P(B|not A) = 0.03125
P(A|B) using Bayes' theorem = 0.6888888888888889


In [18]:
print("P(A|B) direct =", P_A_given_B)
print("P(A|B) using Bayes =", P_A_given_B_bayes)

P(A|B) direct = 0.6888888888888889
P(A|B) using Bayes = 0.6888888888888889


Using Bayes’ theorem, the probability that a student has a part-time job given that the student has a stress score of at least 7 is approximately 68.89%. This agrees exactly with the value obtained using the direct conditional-probability formula, verifying the calculation.

Since P(A ∩ B)=0.05167 is not equal to P(A)P(B)=0.019, the events are not independent. There appears to be an association between having a part-time job and having a high stress score in this dataset.

#Normal Distribution

In [19]:
mean_readiness = df['Academic_Readiness_Score'].mean()
std_readiness = df['Academic_Readiness_Score'].std()

print("Mean =", mean_readiness)
print("Standard Deviation =", std_readiness)

Mean = 71.77316666666667
Standard Deviation = 9.694429667762527


In [20]:
highest_score = df['Academic_Readiness_Score'].max()
lowest_score = df['Academic_Readiness_Score'].min()

print("Highest score:", highest_score)
print("Lowest score:", lowest_score)

Highest score: 98.0
Lowest score: 41.9


In [21]:
z_highest = (highest_score - mean_readiness) / std_readiness

print("Z-score of highest score:", z_highest)

Z-score of highest score: 2.705350828481123


Student's Academic Readiness Score is 2.705 standard deviations above the mean.

In [22]:
z_lowest = (lowest_score - mean_readiness) / std_readiness

print("Z-score of lowest score:", z_lowest)

Z-score of lowest score: -3.081477476287823


Student's Academic Readiness Score is 3.08 standard deviations below the mean.

##68-95-99.7 Rule
Within 1 standard deviation → approximately 68%

Within 2 standard deviations → approximately 95%

Within 3 standard deviations → approximately 99.7%

In [23]:
print("Within 1 SD:", mean_readiness - std_readiness, "to", mean_readiness + std_readiness)

print("\nWithin 2 SD:", mean_readiness - 2*std_readiness, "to", mean_readiness + 2*std_readiness)

print("\nWithin 3 SD:", mean_readiness - 3*std_readiness, "to", mean_readiness + 3*std_readiness)

Within 1 SD: 62.07873699890414 to 81.4675963344292

Within 2 SD: 52.38430733114161 to 91.16202600219172

Within 3 SD: 42.689877663379086 to 100.85645566995424


#Summary
- Academic Readiness Score showed the greatest variability, with a standard deviation of 9.69.
- Daily Screen Time had 18 outliers, indicating some students had unusually high screen usage.
- Academic Readiness ranged from 41.9 to 98.0, showing a wide difference in student preparedness.
- Students with a part-time job had a higher association with high stress; P(A|B) = 68.89%.
- A and B were not independent, since \(P(A\cap B)=0.0517\) differed from \(P(A)P(B)=0.019\).